[c10d] Make broadcast as a custom op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76722

Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

Commit:
[c10d] Make broadcast as a custom op

Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

cbalioglu commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -975,18 +976,21 @@ that adds a prefix to each key inserted to the store.
           .def(
               "broadcast",
-              &::c10d::ProcessGroup::broadcast,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
```
If I understand correctly, this change introduces some discrepancy between the C++ and Python Process Group APIs right? In C++ you have to explicitly call `dispatch_broadcast()` while in Python calling `pg.broadcast()` already goes through dispatcher. Not sure whether we want this sort of behavioral difference. This might also have implication on our existing high-level APIs (i.e. DDP, FSDP, & co.)

ezyang replied:
Yeah you are going to want ProcessGroup::broadcast to also go through the dispatcher, the manual changes below are going to be very error prone otherwise.

alanwaketan replied:
I think we still need to have the real ProcessGroup::broadcast to actually invoke nccl's broadcast. And then ones that route to the dispatcher.

Putting everything under ProcessGroup is an option, but then there are still two sets of APIs. I guess I can having ProcessGroup::broadcast being the dispatcher's one. And then introduce ProcessGroup::broadcast_impl to be the one that calls into nccl.

alanwaketan replied:
@kumpera Could you illustrate what's the benefits by doing so?

wconstab replied:
@alanwaketan can you summarize the outcome of this discussion? was there a conclusion?

alanwaketan replied:
The conclusion is that we have separate APIs for C++ and Python. In Python, the pybind11 is used directly to route to the dispatcher. In C++, since the ProcessGroup's APIs are also the interface for 3rd party to adopt, we can't change that. We have to introduce the ops::xxx API. It's error prone. Given we shouldn't write complicated C++ stuff anymore, it should be okay?

cbalioglu commented on torch/csrc/distributed/c10d/DispatchOps.h:
```diff
@@ -0,0 +1,12 @@
+namespace c10d {
+
+// Belows are essentially ProcessGroup's corresponding ops but routed to the dispatcher.
+TORCH_API c10::intrusive_ptr<ProcessGroup::Work> dispatch_broadcast(const c10::intrusive_ptr<ProcessGroup>& process_group,
```
Instead of calling this `dispatch_broadcast()` which does not really sound much descriptive, I would prefer to have a subnamespace `c10d::ops`, rename this back to `broadcast()` , and move into that namespace.

```
namespace c10d {
namespace ops {

TORCH_API c10::intrusive_ptr<ProcessGroup::Work> broadcast(...);

}
}
```

alanwaketan replied:
I guess using namespace v.s. prefix_ is just a coding style preference. I feel like dispatch_ can really tell the users that this operation will go through the dispatcher.

cbalioglu commented on torch/csrc/distributed/c10d/DispatchOps.cpp:
```diff
@@ -0,0 +1,38 @@
+TORCH_LIBRARY(c10d, m) {
+  // The following ProcessGroup and Work definations are more like declarations.
+  // They don't expose the details of the two classes into TorchScript.
+  m.class_<ProcessGroup>("ProcessGroup")
```
Not sure why we need to define `ProcessGroup` and `Work` Python types here. We already have them defined in `init.cpp`. Am I missing something?

alanwaketan replied:
This is for TorchScript not pybind11.

cbalioglu commented on torch/csrc/distributed/c10d/DispatchOps.cpp:
```diff
@@ -0,0 +1,38 @@
+
+c10::intrusive_ptr<ProcessGroup::Work> dispatch_broadcast(const c10::intrusive_ptr<ProcessGroup>& process_group,
+    at::TensorList tensors, const BroadcastOptions& opts) {
+  auto op = c10::Dispatcher::singleton().findSchemaOrThrow("c10d::broadcast", "")
```
Should we cache the returned `op`? Collective ops such as broadcast can be called pretty frequently in model training. Not sure how much price we pay each time we call `findSchemaOrThrow()`.

ezyang replied:
conventionally this is marked static so it's cached. Should be done so here

cbalioglu reviewed (commented):
Left a couple comments. I also don't see any accompanying tests. Do you plan to have any?

aazzolini commented:
2 things:

1) I believe exposing ProcessGroup at this level is not the right abstraction since it's more related to the runtime than it is to the mathematical operation being performed. I think for things like LazyTensor and MetaTensor specifically , process group doesn't make sense at all.

We will need to distill the semantic-carrying part of ProcessGroup (e.g. device ids / ranks, for example) from the "runtime-side" part of it. 

My suggestion would be to just pass a list of device IDs' just like JAX does. As for the "concrete" process group to pick from, I believe this is more related to the "device" where this runs. Is it possible to come up with a Device that holds a process group?

2) I believe we should not expose "Work" either since pytorch already has the concept of std:Future. Could we wrap the work in a future and return it?

alanwaketan commented:
> 2 things:
> 
> 1. I believe exposing ProcessGroup at this level is not the right abstraction since it's more related to the runtime than it is to the mathematical operation being performed. I think for things like LazyTensor and MetaTensor specifically , process group doesn't make sense at all.
> 
> We will need to distill the semantic-carrying part of ProcessGroup (e.g. device ids / ranks, for example) from the "runtime-side" part of it.
> 
> My suggestion would be to just pass a list of device IDs' just like JAX does. As for the "concrete" process group to pick from, I believe this is more related to the "device" where this runs. Is it possible to come up with a Device that holds a process group?
> 
> 2. I believe we should not expose "Work" either since pytorch already has the concept of std:Future. Could we wrap the work in a future and return it?

Should we leave this to @cbalioglu's redesign of c10d ops?

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
> Left a couple comments. I also don't see any accompanying tests. Do you plan to have any?

Thanks for reviewing the patch. I think existing tests should have covered this pretty well. I will have separate PRs for E2E integration examples, which should add more coverages.

cbalioglu commented:
> Should we leave this to @cbalioglu's redesign of c10d ops?

There are going to be some organizational changes in our team starting next week. So most likely this task will be assigned to someone else, but I will keep you posted.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -203,16 +203,13 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
-        c10::str(
-            "ProcessGroup ", getBackendName(), "does not support broadcast"));
-  }
+  c10::intrusive_ptr<ProcessGroup::Work> broadcast(
```
This method needs to be virtual otherwise there might be issues in terms of appropriate dispatching when derived classes override this method.

alanwaketan replied:
I don't think we want the derived classes to override this method. This is the key entry point to the dispatcher. Will it make more sense to separate the dispatch passable cc ops into a separate file and then route the pybind and any C++ distributed runtime, for instance, reducer to call into the new cc ops?

But then we need a way to make sure people don't accidentally use the old ops. So that's why I think it makes more senses to directly modify the existing ops here.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -5,6 +5,28 @@
 namespace c10d {
+namespace {
+
+c10::intrusive_ptr<ProcessGroup::Work> broadcast(const c10::intrusive_ptr<ProcessGroup>& process_group,
+    at::TensorList tensors, int64_t root_rank, int64_t root_tensor, int64_t timeout) {
```
Why is the signature of this broadcast method different from https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1162?

alanwaketan replied:
The parameters are unbundled from BroadcastOptions. `dist.broadcast` is a subset of this API.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -929,7 +929,7 @@ class AsyncBroadcastCUDAWork : public AsyncBroadcastWork {
 } // namespace
 
-c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast_impl(
```
Our solution can't rely on changing all subclasses since there are third-party implementations of the ProcessGroup interface that we don't control.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -186,4 +208,15 @@ void ProcessGroup::init() {
   C10_LOG_API_USAGE_ONCE(fmt::format("c10d.process_group_{}", getBackendName()));
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroup::broadcast(
+    std::vector<at::Tensor>& tensors, const BroadcastOptions& opts) {
```
So the python API here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1162 does additional validation before calling the ProcessGroup.

Wouldn't this lead to some semantic differences where the same thing run in eager would have additional validation and might fail, whereas when you trace the op it doesn't have the additional validation and it might run into a weird error downstream instead?

Shouldn't we modify the python API to just directly call into this so that both eager and traced version have same behavior?

alanwaketan replied:
This is a good question. During the first trace, the tracer will make sure that the inputs can pass those checks. During the replay, the tracer will make sure the subsequence inputs share critical properties as the first trace. Otherwise, it will start a new trace.

pritamdamania87 commented:
@alanwaketan One more thing I just realized was about autograd. The aten ops probably also need to include a backward op as well for the collectives. Our collectives ops unfortunately were not autograd enabled by default when they were built out, but for these aten ops, we should have them be autograd enabled.

For BC purposes we probably would still have to have our public c10d APIs call something like `torch.no_grad` and then call the aten ops.

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

kumpera reviewed (commented):
Could the dispatch-able version take a PG-like object that lacks any methods?

Wouldn't we need something like that anyways if we want to enable models with collectives to run in a local fashion and possibly without C10D compiled in?

As we go this route, we should make those collectives return Future and not Work.

alanwaketan commented:
> Could the dispatch-able version take a PG-like object that lacks any methods?
> 

Could you elaborate more on how that's going to work? How this wrapper can call the actual implementation then?

> Wouldn't we need something like that anyways if we want to enable models with collectives to run in a local fashion and possibly without C10D compiled in?
> 

Could you elaborate more what's local fashion?

> As we go this route, we should make those collectives return Future and not Work.

Replacing Work with Future should be a follow up given Work is deeply integrated within the actual implementation. Unless you are suggesting we just extract the Future from the Work and return, which I don't think it's a good idea.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -203,8 +203,13 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
 
   // Subclasses must override this method to return the backend name
-  virtual const std::string getBackendName() const = 0;
+  virtual const std::string getBackendName() const {
+    TORCH_INTERNAL_ASSERT(false, "getBackendName is not implemented.");
+  };
```
Why do we need this change?

alanwaketan replied:
Just when registering a class/type with TorchScript, it doesn't like pure virtual method in the class.

pritamdamania87 replied:
cc @suo Is there a way to fix this? It doesn't seem right that we have to change a core interface method like this from pure virtual for TorchScript.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -20,7 +21,7 @@ class BroadcastWork {
     BroadcastOptions broadcastOptions;
     broadcastOptions.rootRank = root_rank;
-    work_ = process_group->broadcast(flat_tensor_, broadcastOptions);
+    work_ = ops::broadcast(process_group, flat_tensor_, broadcastOptions);
```
I think for DDP, we should just register a `_broadcast_coalesced` op and directly call that from python here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/utils.py#L131. Can be done in a separate PR.

pritamdamania87 replied:
I feel it is cleaner for python to just call into an aten op. In this case, DDP is calling _broadcast_coalesced and there is a bunch of python and C++ logic before we actually call into the `broadcast` cc op.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2275,8 +2275,9 @@ def call_dist_op(
+                # We are only interested in the backend's implementation not the dispatcher wrapper.
                 events = get_profiling_event(
-                    profiling_title_postfix, autograd_profiler_ctx
+                    dist.get_backend()+profiling_title_postfix, autograd_profiler_ctx
```
Are these changes needed?

alanwaketan replied:
If we don't add the backend prefix, the events returned will be the nccl::broadcast and newly added broadcast op. I think the intention here is to only capture the nccl::broadcast.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -975,18 +976,21 @@ that adds a prefix to each key inserted to the store.
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
+                 const std::vector<at::Tensor>& tensors,
+                 ::c10d::BroadcastOptions opts) {
+                return ::c10d::ops::broadcast(self, tensors, opts);
```
This does capture all the broadcast calls from the python side, however I think the ideal place is to call the ops::broadcast directly here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1182. I'm worried that the traced version of the program won't have the python validation? For example, there is this logic on the python side that is important:

```
    if _rank_not_in_group(group):
        _warn_not_in_group("broadcast")
        return
```

I guess the python side of things would still run with something like TorchDynamo, but if there is something that traces a program in a language agnostic IR and would like to run the IR in C++, we would run into issues.

This can be done in a follow up PR since it is probably a much bigger change, but I feel our python API should directly call the aten op and not have any python logic before that.

wconstab replied:
> I'm worried that the traced version of the program won't have the python validation?

This depends on the tracer. For dynamo, we should ensure that it is replacing the python validation with its own guards.  (That may actually not be the case today as dynamo mostly has special logic for tensors/nn.modules and friends.

If we don't use dynamo, e.g. for a direct AotAutograd we'd need another policy for how to ensure safety, but it wouldn't happen by default.

pritamdamania87 reviewed (commented):
Looks good overall, have a few comments inline. I'm assuming follow up PRs would add autograd support etc?

ezyang commented:
is this ready for wider review?

alanwaketan commented:
@pritamdamania87 Thanks for reviewing the PR. I have summarized the requested follow ups as follows:
1. Add autograd support, which can probably reuse the implementation here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/nn/functional.py.
2. Ensure the guards in distributed_c10d.py wrappers get executed or reassured in the replay of the captured graph in any graph mode.
3. Potentially add a new `_broadcast_coalesced` op for DDP.

My question is can I land this PR now and apply this technique to other cc ops before the above follow ups? I think it's more important to route all cc ops to the dispatcher before the above follow ups given a lot of other works relies on it.

pritamdamania87 reviewed (approved):
@alanwaketan The follow ups look good to me! Would be great if we can create some gh issues to track this work as well. The changes look good to me, however please check with @ezyang to see if he is comfortable with landing this PR.

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This patch makes broadcast as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops
python test/distributed/test_c10d_gloo.py -k test_broadcast_basics
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
> @alanwaketan The follow ups look good to me! Would be great if we can create some gh issues to track this work as well. The changes look good to me, however please check with @ezyang to see if he is comfortable with landing this PR.

Opened three issues to keep track of the follow ups here: #79467, #79468, #79469.

wconstab commented on torch/csrc/distributed/c10d/Ops.hpp:
```diff
@@ -0,0 +1,14 @@
+namespace c10d {
+namespace ops {
+
+// Belows are essentially ProcessGroup's corresponding ops but routed to the dispatcher.
```
typo 'Below'

alanwaketan replied:
Oops, didn't spot this and let me fix that in my next PR.

alanwaketan commented:
@pytorchbot merge --green
